framework/src/preconditioners/DualMortarPreconditioner.C

```diff
@@ -0,0 +1,588 @@
+  // to save memory, no new matrix is created
+  for (unsigned int i=0; i<_D->m(); ++i)
+  {
+   Number tmp = 1.0/(*_D)(i,i);
+   std::vector<numeric_index_type> row_i;
+   row_i.push_back(i);
+    _D->zero_rows(row_i, tmp);
+  }
+  _D->close();
+
+
+  // compute MDinv=_M*_D
+  _M->matrix_matrix_mult(*_D, *_MDinv); // (should use empty initializer for _MDinv)
+  _MDinv->close();
+
+  // initialize _J_condensed
+
+  _matrix->create_submatrix(* _J_condensed, _rows, _cols);
+  std::unique_ptr<PetscMatrix<Number>> J_trial(libmesh_make_unique<PetscMatrix<Number>>(MoosePreconditioner::_communicator));
+  _matrix->create_submatrix(* J_trial, _rows, _cols);
+
+  // compute changed parts: MDinv*K2ci, MDinv*K2cc
+  std::unique_ptr<PetscMatrix<Number>> MDinvK2ci(libmesh_make_unique<PetscMatrix<Number>>(MoosePreconditioner::_communicator)),
+  MDinvK2cc(libmesh_make_unique<PetscMatrix<Number>>(MoosePreconditioner::_communicator));
+  _matrix->create_submatrix( *MDinvK2ci, u1c, u2i); // get MDinvK2ci initialized (should use empty initializer here)
+  _matrix->create_submatrix( *MDinvK2cc, u1c, u2c); // get MDinvK2cc initialized (should use empty initializer here)
+  _MDinv->matrix_matrix_mult(*_K2ci, *MDinvK2ci);
+  _MDinv->matrix_matrix_mult(*_K2cc, *MDinvK2cc);
+
+  MDinvK2ci->close();
+  MDinvK2cc->close();
+
+  // add changed parts to _J_condensed
+  // original system row_id: u1c
+  // original system col_id: u2i, u2c
+  std::vector<numeric_index_type> row_id_cond, col_id_cond_u2i, col_id_cond_u2c;
+  std::map<numeric_index_type, numeric_index_type> row_id_cond_mp, col_id_cond_u2i_mp, col_id_cond_u2c_mp;
+
+  for (auto it: index_range(u1c))
+  {
+    numeric_index_type lid=static_cast<numeric_index_type>(it);
+    auto it_row = find (_rows.begin(), _rows.end(), u1c[it]);
+    if (it_row != _rows.end())
+    {
+      numeric_index_type gid = std::distance(_rows.begin(), it_row);
+      row_id_cond_mp.insert(std::make_pair(lid , gid));
+    }
+    else
+      mooseError("DOF ", u1c[it]," does not exist in the rows of the condensed system");
+  }
+
+  for (auto it: index_range(u2i))
+  {
+    numeric_index_type lid=static_cast<numeric_index_type>(it);
+    auto it_col = find(_cols.begin(), _cols.end(), u2i[it]);
+    if (it_col != _cols.end())
+    {
+      numeric_index_type gid = std::distance(_cols.begin(), it_col);
+      col_id_cond_u2i_mp.insert(std::make_pair(lid, gid));
+    }
+    else
+      mooseError("DOF ", u2i[it]," does not exist in the columns of the condensed system");
+  }
+
+  for (auto it: index_range(u2c))
+  {
+    numeric_index_type lid=static_cast<numeric_index_type>(it);
+    auto it_col = find(_cols.begin(), _cols.end(), u2c[it]);
+    if (it_col != _cols.end())
+    {
+      numeric_index_type gid = std::distance(_cols.begin(), it_col);
+      col_id_cond_u2c_mp.insert(std::make_pair(lid, gid));
+    }
+    else
+      mooseError("DOF ", u2c[it]," does not exist in the columns of the condensed system");
+  }
+
```

Comment:
We need to somehow preallocate `_J_condensed ` here.

My thought is: 

1) compare the sparsity of `MDinvK2ci`  with that of `K2ci`,  

2) Retrieve the sparsity of `_J_condensed` by looping over rows.

3) Appending  differences, we recreate a matrix `_J_condensed_new`

4) `_J_condensed_new-> add_sparse_matrix(_J_condensed)`

Reply:
Would you mind pointing to me where the matrix sparsity information are? Seems to be an important component here.